### PR TITLE
Added tests for Ipcsocket

### PR DESCRIPTION
**Pull request by awelling2801**
**Work item:** _"Internal"

**What were the changes?**  
Added Unit tests to increase code coverage for ipcsocket.cc
This PR adds one file - test/IpcsocketTests.cpp comprising of 5 tests.

**Why were the changes made?**  
The test improved code coverage for ipcsocket.cc
|| Function Coverage | Line Coverage | Region coverage | Branch Coverage
|--|-- | -- | -- | --
Default Coverage | 42.86% (3/7) |   24.68% (39/158) |   22.36% (53/237) |   18.29% (15/82)
Updated Coverage |  100.00% (7/7) |   77.71% (122/157) |   58.09% (79/136) |   35.00% (28/80)

**How was the outcome achieved?**  
The outcome was achieved by adding a IpcsocketTests.cpp file comprising of the tests.
The visibility set to default instead of hidden achieved by https://github.com/ROCm/rccl/pull/1664
Also the code coverage was run on the Unit test binary and not librccl.so
The environment variables used were as follows - 

`NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL RCCL_ENABLE_INTRANET=1 NCCL_DMABUF_ENABLE=1 RCCL_GDR_FLUSH_GPU_MEM_NO_RELAXED_ORDERING=0 ./rccl-UnitTests --gtest_filter=Ipcsocket.SendAndReceiveFd:Ipcsocket.SocketInitNullHandle:Ipcsocket.SocketGetFdNullHandle:Ipcsocket.SocketCloseNullHandle:Ipcsocket.SocketCloseNegativeHandle `

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

File: test/IpcsocketTests.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+#include "TestBed.hpp"
+
+namespace RcclUnitTesting
+{
+  TEST(Ipcsocket, SocketInitNullHandle){
+    int rank = 0;
+    uint64_t hash = 0x1234;
+    volatile uint32_t abortFlag = 0;
+    // Call function with NULL handle
+    ncclResult_t result = ncclIpcSocketInit(NULL, rank, hash, &abortFlag);
```
Comment (corey-derochie-amd) [inline test/IpcsocketTests.cpp:22]: These symbols are not exported from librccl, so compilation fails.

**Review by corey-derochie-amd — DISMISSED** [review]

File: test/IpcsocketTests.cpp (hunk 1 above)
Comment (corey-derochie-amd) [reply test/IpcsocketTests.cpp:22]: It may be necessary to link ipcsocket.cc directly into rccl-UnitTests to test this unit.

Comment (awelling2801) [reply test/IpcsocketTests.cpp:22]: This commit follows - https://github.com/ROCm/rccl/pull/1664. Which sets the visibility to default instead of hidden to expose internal library symbols for unit testing the RCCL library. 
If you are testing this for now - I would suggest changing the visibility to default in CMakeLists.txt

**Comment by venksubramd** [comment]
The test for ncclIpcSocketInit is checking if the returned error code for handle being NULL is correct. This tests the initial check done by the ncclIpcSocketInit function. However, we do not have any tests for what the ncclIpcSocketInit actually is supposed to do. What is the intent of ncclIpcSocketInit, what is it supposed to achieve, and what are the possible failures it may encounter beyond the handle being NULL? We would need a lot more tests for that, beyond the one test we currently have.


The test for ncclIpcSocketGetFd is checking if the returned error code for handle being NULL is correct. We would use at least two more tests to check what the function should do if the handle is not NULL but for different values for the fd argument.

We have two tests for ncclIpcSocketClose, the first checks the behavior if the handle is NULL. The second test checks the behavior if the handle's fd is less than zero. We could use a few more tests for the two steps after the initial few checks in this function. 

We have one test for the ncclIpcSocketRecvFd function. This function, however, has only one line where it calls the ncclIpcSocketRecvMsg. It may be better to write tests for the ncclIpcSocketRecvMsg function instead of ncclIpcSocketRecvFd or may be a very small test for ncclIpcSocketRecvFd that merely checks that this function calls ncclIpcSocketRecvMsg with proper arguments, by mocking ncclIpcSocketRecvMsg. This one test is quite long and has a if-else part. Generally we should avoid branches in unit tests as we have no control, in that case, over running all paths in a single run of the test.

We don't have any tests for ncclIpcSocketSendFd and ncclIpcSocketSendMsg?

**Comment by awelling2801** [comment]
>  However, we do not have any tests for what the ncclIpcSocketInit actually is supposed to do. What is the intent of ncclIpcSocketInit, what is it supposed to achieve, and what are the possible failures it may encounter beyond the handle being NULL?

These scenarios are already exercised by other test cases. It would be helpful to review the existing code coverage first to avoid redundant test efforts.

There cases that need mocking for system level functions need their own PR.

>  We would use at least two more tests to check what the function should do if the handle is not NULL but for different values for the fd argument. 

Already covered by other test cases and do not need their own tests 

> We have two tests for ncclIpcSocketClose, the first checks the behavior if the handle is NULL. The second test checks the behavior if the handle's fd is less than zero. We could use a few more tests for the two steps after the initial few checks in this function.

The case for - 
`#ifndef USE_ABSTRACT_SOCKET
  if (handle->socketName[0] != '\0') {
    unlink(handle->socketName);
  }
#endif` 
The unlink() block is conditionally compiled only when USE_ABSTRACT_SOCKET is not defined. In our current build setup, USE_ABSTRACT_SOCKET is defined, so this section is omitted entirely during compilation. Hence we do not need tests for it.
The rest of the path is already covered by other tests and we do not need redundancy. 

> It may be better to write tests for the ncclIpcSocketRecvMsg function instead of ncclIpcSocketRecvFd or may be a very small test for ncclIpcSocketRecvFd that merely checks that this function calls ncclIpcSocketRecvMsg with proper arguments, by mocking ncclIpcSocketRecvMsg. 

Not sure why this is necessary. The ncclIpcSocketRecvFd does nothing except call ncclIpcSocketRecvMsg. 

> We don't have any tests for ncclIpcSocketSendFd and ncclIpcSocketSendMsg?

Tests for both send and receive. Not really possible to do one without the other and this was after discussion with the team

**Comment by venksubramd** [comment]
@awelling2801 

Could you please create a new PR for the tasks that will require mocking so we can focus on that effort separately. For the parts that are tested using other tests, could you please check if those tests actually examine the behavior (which is necessary) or there merely focus on code coverage (which is not sufficient by itself).

**Comment by atulkulk** [comment]
> @awelling2801
> 
> Could you please create a new PR for the tasks that will require mocking so we can focus on that effort separately. For the parts that are tested using other tests, could you please check if those tests actually examine the behavior (which is necessary) or there merely focus on code coverage (which is not sufficient by itself).

Created ticket for additional work: [ GPUAI-5988 ]

**Comment by venksubramd** [comment]
@awelling2801 @atulkulk  Thank you for creating ticket for the additional work. We can close this one.

**Review by atulkulk — APPROVED** [review]

**Review by corey-derochie-amd — APPROVED** [review]
